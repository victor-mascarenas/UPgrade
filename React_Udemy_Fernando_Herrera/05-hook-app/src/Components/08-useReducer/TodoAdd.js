import React, { useRef } from 'react';
import useForm from '../../hooks/useForm';

export const TodoAdd = ({addTodo}) => {
    const [{description}, inputOnChange, reset] = useForm({
        description: ''
    });
    const input = useRef();

    const formOnSubmit = (e) => {
        e.preventDefault();
        if (description.trim().length <= 1) {
            return;
        }
        addTodo({
            id: new Date().getTime(),
            desc: description,
            done: false
        });
        reset();
        input.current.focus();
    };

    return (
        <>
            <h4>Agregar ToDo</h4>
            <hr/>
            <form onSubmit={formOnSubmit}>
                <input type='text' name='description' placeholder='Comprar...' autoComplete='false' className='form-control' onChange={inputOnChange} value={description} ref={input}/>
                <button className='btn btn-outline-primary mt-1 col-12' type='submit'>Agregar</button>
            </form>
        </>
    );
};
