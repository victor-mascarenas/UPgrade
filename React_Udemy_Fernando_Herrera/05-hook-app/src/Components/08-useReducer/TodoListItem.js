import React from 'react';

export const TodoListItem = ({todo, i: index, pOnClick, buttonOnClick}) => {
    return (
        <li key={todo.id} className='list-group-item'>
            <p onClick={pOnClick.bind(null, todo.id)} className={todo.done ? 'completed' : ''}>
                {index + 1}. {todo.desc}
            </p>
            <button className='btn btn-danger' onClick={buttonOnClick.bind(null, todo.id)}>Borrar</button>
        </li>
    );
};
