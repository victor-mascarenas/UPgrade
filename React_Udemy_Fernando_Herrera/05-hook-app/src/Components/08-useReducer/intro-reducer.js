const todosInitial = [{//Estado inicial
    id: 1,
    todo: 'Comprar pan',
    done: false
}];

const todoReducer = (state = todosInitial, action) => {
    if (action) {
        if (action.type === 'agregar') {
            state = [...state, action.payload];
        }
    }
    return state;
};

let todos = todoReducer();

const newTodo = {
    id: 2,
    todo: 'Comprar leche',
    done: false
};
const action = {
    type: 'agregar',
    payload: newTodo
};
todos = todoReducer(todos, action);

console.log(todos);