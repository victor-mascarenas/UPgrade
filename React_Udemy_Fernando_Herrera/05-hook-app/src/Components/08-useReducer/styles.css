body {
    padding: 50px;
}
p {
    cursor: pointer;
    margin: 0px;
}

.complete {
    text-decoration: line-through;
}
.list-group-item {
    align-items: center;
    display: flex;
    justify-content: space-between;
}