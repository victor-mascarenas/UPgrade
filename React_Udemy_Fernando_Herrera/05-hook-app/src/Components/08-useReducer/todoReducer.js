//Delete/toggle: paylod = id
export const todoReducer = (state, action) => {
    if (action) {
        switch (action.type) {
            case 'add':
                state = [...state, action.payload];
                break;
            case 'delete':
                state = state.filter((todo) => todo.id !== action.payload);
                break;
            case 'toggle':
                state = state.map(todo =>
                    (todo.id === action.payload)
                        ? {...todo, done: !todo.done}
                        : todo
                );
                break;
            default:
                break;
        }
    }
    return state;
};
