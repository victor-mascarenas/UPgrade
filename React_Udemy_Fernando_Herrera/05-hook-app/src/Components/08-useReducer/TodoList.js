import React from 'react';
import { TodoListItem } from './TodoListItem';

export const TodoList = ({todos, pOnClick, buttonOnClick}) => {
    return (
        <ul className='list-group list-group-flush'>
            {
                todos.map((todo, i) => {
                    return <TodoListItem key={todo.id} todo={todo} i={i} pOnClick={pOnClick} buttonOnClick={buttonOnClick}/>;
                })
            }
        </ul>
    );
};
