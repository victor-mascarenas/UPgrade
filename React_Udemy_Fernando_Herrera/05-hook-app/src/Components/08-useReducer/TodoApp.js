import React, { useEffect, useReducer, useRef } from 'react';
import './styles.css';
import { todoReducer } from './todoReducer';
import useForm from '../../hooks/useForm';
import { TodoList } from './TodoList';

const init = () => {
    return JSON.parse(localStorage.getItem('todos')) || [];
};

export const TodoApp = () => {
    const [todos, dispatch]= useReducer(todoReducer, [], init);
    const [{description}, inputOnChange, reset] = useForm({
        description: ''
    });
    const input = useRef();
    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);

    const formOnSubmit = (e) => {
        e.preventDefault();
        if (description.trim().length <= 1) {
            return;
        }
        dispatch({
            type: 'add',
            payload: {
                id: new Date().getTime(),
                desc: description,
                done: false
            }
        });
        reset();
        input.current.focus();
    };
    const buttonOnClick = (id) => {
        dispatch({
            type: 'delete',
            payload: id
        });
    };
    const pOnClick = (id) => {
        dispatch({
            type: 'toggle',
            payload: id
        });
    };

    return (
        <div>
            <h1>Todo App ({todos.length})</h1>
            <hr/>
            <div className='row'>
                <div className='col-7'>
                    <TodoList todos={todos} pOnClick={pOnClick} buttonOnClick={buttonOnClick}/>
                </div>
                <div className='col-5'>
                    <h4>Agregar ToDo</h4>
                    <hr/>
                    <form onSubmit={formOnSubmit}>
                        <input type='text' name='description' placeholder='Comprar...' autoComplete='false' className='form-control' onChange={inputOnChange} value={description} ref={input}/>
                        <button className='btn btn-outline-primary mt-1 col-12' type='submit'>Agregar</button>
                    </form>
                </div>
            </div>
        </div>
    );
};
