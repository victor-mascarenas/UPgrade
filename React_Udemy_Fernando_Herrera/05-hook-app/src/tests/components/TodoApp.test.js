import React from 'react';
import {shallow, mount} from 'enzyme';
import { TodoApp } from '../../Components/08-useReducer/TodoApp';
import { demoTodos } from '../fixtures/demoTodos';
import { act } from '@testing-library/react';

describe('Pruebas en TodoApp', () => {
    const wrapper = shallow(<TodoApp/>);
    Storage.prototype.setItem = jest.fn(() => {});

    test('Debe mostrarse correctamente', () => {
        expect(wrapper).toMatchSnapshot();
    });
    test(`Debe de agregar dos todo's`, () => {
        const wrapper = mount(<TodoApp/>);
        act(() => {
            wrapper.find('TodoAdd').prop('addTodo') (demoTodos[0]);
            wrapper.find('TodoAdd').prop('addTodo') (demoTodos[1]);
        });
        expect(wrapper.find('h1').text().trim()).toBe(`Todo App (${demoTodos.length})`);
        expect(localStorage.setItem).toHaveBeenCalledTimes(2);
    });
    test('Debe eliinar un todo', () => {
        wrapper.find('TodoAdd').prop('addTodo') (demoTodos[0]);
        wrapper.find('TodoList').prop('buttonOnClick') (demoTodos[0].id);
        expect(wrapper.find('h1').text().trim()).toBe(`Todo App (0)`);
    });
    
});
