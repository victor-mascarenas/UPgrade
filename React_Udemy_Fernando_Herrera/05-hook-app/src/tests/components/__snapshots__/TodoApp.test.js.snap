// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en TodoApp Debe mostrarse correctamente 1`] = `
<div>
  <h1>
    Todo App (
    0
    )
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-7"
    >
      <TodoList
        buttonOnClick={[Function]}
        pOnClick={[Function]}
        todos={Array []}
      />
    </div>
    <div
      className="col-5"
    >
      <TodoAdd
        addTodo={[Function]}
      />
    </div>
  </div>
</div>
`;
